src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs

```diff
@@ -0,0 +1,175 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NLog;
+using SFA.DAS.Payments.DCFS.Domain;
+using SFA.DAS.ProviderPayments.Calc.Refunds.Domain;
+using SFA.DAS.ProviderPayments.Calc.Refunds.Services.Dependiencies;
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.ProviderPayments.Calc.Refunds.Services
+{
+    public class LearnerRefundProcessor : IProcessLearnerRefunds
+    {
+        private readonly ILogger _logger;
+
+        private static readonly List<FundingSource> FundingSourcesToRefund = new List<FundingSource>
+        {
+            FundingSource.Levy,
+            FundingSource.CoInvestedEmployer,
+            FundingSource.CoInvestedSfa,
+            FundingSource.FullyFundedSfa
+        };
+
+        public LearnerRefundProcessor(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public List<PaymentEntity> ProcessRefundsForLearner(
+            List<RequiredPaymentEntity> refunds,
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundGroups = refunds.ToLookup(x => new RefundGroup(x));
+            var volatilePreviousPayments = new List<HistoricalPaymentEntity>(previousPayments);
+
+            var refundPayments = new List<PaymentEntity>();
+            
+            foreach (var refundGroup in refundGroups)
+            {
+                foreach (var refund in refundGroup)
+                {
+                    var previousPaymentGroup = volatilePreviousPayments.ToLookup(x => new RefundGroup(x));
+                    var previousPaymentsForRefundGroup = new List<HistoricalPaymentEntity>();
+                    try
+                    {
+                        if (previousPaymentGroup.Contains(refundGroup.Key))
+                        {
+                            previousPaymentsForRefundGroup = previousPaymentGroup[refundGroup.Key].ToList();
+                        }
+
+                        var newRefunds = ProcessRefund(refund, previousPaymentsForRefundGroup);
+                        refundPayments.AddRange(newRefunds);
+                        volatilePreviousPayments.AddRange(newRefunds.Select(x => new HistoricalPaymentEntity(x, refund)));
+                    }
+                    catch(ApplicationException)
```

Comment:
We are using this exception for a very particular case, ApplicationException Is invalid and it makes sense to create a custom exception (if we decide to do this in this way)